Metadata support for ImageSource class.

vineetbansal opened the pull request:
A `metadata` DataFrame is attached to all `ImageSource` objects. 'Special' fields in this metadata structure are used to keep track of amplitudes/offsets/rotation angles/filters etc. These special fields can be set/get using regular attribute access on the `ImageSource` class. `get_metadata` and `set_metadata` methods can be used for other fields in the future.

When an `ImageSource` is converted to a `Starfile` object (in preparation for serializing to a *.star file), these 'special' fields can take on aliases (For example `RelionStarfileStack` subclasses `StarfileStack` and renames `offset_x` as `_rlnOriginX`.

Note that metadata fields can, and do support, arbitrary object storage, as is done to store `Filter` objects. This storage in numpy is through pointers to actual objects, so the objects themselves are not duplicated.

The `SourceFilter` class (which essentially just kept track of the association between `Filter` objects and `ImageSource` objects by maintaining indices) is no longer needed and thus removed. For now, callers who set `Filter` objects on a Simulation can keep track of the indices if they want (though they shouldn't have to). We can add support methods for this stuff if needed.

Other changes include a `Starfile` object to keep track of records obtained from a *.star file (DataFrames are used internally to keep track of loop data, key value data is supported too), and smarter parsing of huge starfiles by incrementally reading data till `max_rows` is satisfied.

vineetbansal committed:
Metadata support for ImageSource class.

A metadata DataFrame is attached to all ImageSource objects. 'Special' fields in this metadata structure are used to keep track of amplitude/offset/rotation angle/filter etc. These special fields can be set/get using regular attribute access on the ImageSource class. Note that metadata fields can, and do, support arbitrary object storage, as is done to store Filter objects. The SourceFilter class is no longer needed and thus removed. Other changes include a Starfile object to keep track of records obtained from a starfile (DataFrames are used internally to keep track of loop data), and smarter parsing of huge starfiles by incrementally reading data till max_rows is satisfied.

vineetbansal committed:
Equality checks for Starfile* classes.

Added unit test to check serialization/equality of serialized object.

vineetbansal committed:
Renamed COPYING to LICENSE to keep things simple

No real use of a custom LICENSE.md. Also allows GitHub infrastructure to pick up license info automatically.

vineetbansal committed:
minor bug fixes

vineetbansal committed:
Removed intermediate StarfileStack class

Simplified hierarchy - RelionStarfileStack now directly subclasses ImageSource (encountering Starfiles that are not Relion Starfiles deemed very unlikely).

vineetbansal committed:
im_* functions moved as methods inside Image class.

_im_translate/_im_translate2 need to be handled. Added a sample notebook on using the Image class.

janden commented on examples/denoise_covar2d.py:
```diff
@@ -51,12 +50,15 @@
 
-# Create a simulation object
+# Create filters
+filters = [RadialCTFFilter(pixel_size, voltage,defocus=d, Cs=2.0, alpha=0.1)
```
Space after comma.

vineetbansal replied:
fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/source/__init__.py:
```diff
@@ -115,7 +168,31 @@ def _images(self, start=0, num=None):
+
+        return im
+
+    def eval_filter_grid(self, L):
```
Why should this be a method attached to an ImageSource? Doesn't it make more sense to have the filters evaluate *themselves* on a grid? What do we gain by having this here?

vineetbansal replied:
Yes .. I added an issue #38 for this. Let's keep track of it in the Issue tracker instead of right here (will get messy to track along with everything else, and removing a method will generate several cascades of changes throughout the calling code too).

janden commented on examples/denoise_covar2d.py:
```diff
@@ -77,23 +79,21 @@
 # Assign the CTF information and index for each image
-h_idx = sim.filters.indices
-filters = sim.filters.filters
+h_idx = np.array([filters.index(f) for f in sim.filters])
```
This seems a bit awkward (as we discussed). Is there no better way to do this that avoid explicitly indexing the filters?

vineetbansal replied:
There's no need for any *internal* code within `aspire` to keep track of indices. Not sure why this example script needs it. My suspicion is that whatever `h_idx` is being used for, it can be done away with. Something to bring up with Junchao next week.

junchaoxia replied:
For cov2d denoising, the `h_idx` is needed to track the ctf filter for each imag because the mean and covariance values are computed by the ctf groups.  In experimental dataset, the `h_idx` information can be read in as metadata of images.

janden replied:
Couldn't the CTF groups be extracted by working with the `filters` array instead of these proxy indices?

junchaoxia replied:
In Vineet latest change, the `filters.index(f)` gets the index of filter `f`. It might be more reasonable to include an indices as a metadata in the `Image` object if we decide to include all metadata in the `Image` class. In the real experimental dataset, the indices to the CTF filters should be built when read in the star file from the CTF finding process.

vineetbansal replied:
I'm closing this here. An upcoming `IndexedTransformer` or something similar will allow the user to attaching one or more `Transformer` objects in a `Pipeline` attached to an `ImageSource`, and we'll add a method there to return the index (or indices) of any `Transformer` that the caller provides.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+
+        return Image(im_ds)
+
+    def filter(self, filter=None, filter_values=None):
```
Why allow `filter_values` here? Seems to be a cleaner approach to just accept `filter`s.

vineetbansal replied:
I brought this up with Junchao last week.
https://github.com/ComputationalCryoEM/ASPIRE-Python/pull/32#discussion_r314874329

Not sure I understood why he needed this. Will bring this up next week.

junchaoxia replied:
The computing is less if we pass the filters as precomputed values in 2D matrix since the number of filter types should be much smaller as the number of particle images.

vineetbansal replied:
If the caller (`denoise_covar2d.py` and the corresponding test script) simply used `sim.eval_filters` by passing in an Image object (in your case, the clean images), instead of first using `sim.eval_filter_grid` (which should be removed), then `eval_filters` groups processing by unique filters anyway, and calls the above `filter` method without having to worry about the `filter_values` parameter.

vineetbansal replied:
Fixed in e85a5619bd1ea7bd8f5cac427b61f8f881f00e24. A separate issue can be opened in case this is unsatisfactory and there's something I'm missing.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
     def rotate(self):
         raise NotImplementedError
 
+    def save(self, mrcs_filepath, overwrite=False):
```
Not sure how closely we want to tie the `Image` class into the cryo-EM functionality. Ideally, this would be a general-purpose image-processing class that we could use in other settings as well. It might make more sense for this to be outside (in the MRC I/O module, for example).

junchaoxia replied:
Probably a generalized output function is enough. In the derived class it can be overrided in order to save to mrc files. I think that capability of outputting to new mrc files is import to save the processed images such as denoising or downsampling.

vineetbansal replied:
Opened a new issue #48 for this. Closing here.

janden commented on examples/denoise_covar2d.py:
```diff
@@ -77,23 +79,21 @@
 sim.cache(imgs_ctf_clean)
 
+# imgs_ctf_clean is an Image object. Convert to numpy array for subsequent statements
+imgs_ctf_clean = imgs_ctf_clean[:]
```
Is this the standard way to obtain a numpy array from an image? It seems that there's also the `asnumpy()` method that might be used for the purpose. If so, that would be more explicit.

vineetbansal replied:
Yes .. this one slipped through when I was converting all `[:]`s to `.asnumpy()`. Will fix.

vineetbansal replied:
Fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/aspire/abinitio.py:
```diff
@@ -13,6 +13,7 @@
 from aspire.exceptions import WrongInput
 from aspire.aspire.common.logger import logger
 from aspire.aspire.utils.data_utils import validate_square_projections, c_to_fortran
+from aspire.image import _im_translate2 as im_translate
```
I'm assuming this is a temporary fix that will be resolved before merging?

vineetbansal replied:
Yes .. issue #36 will keep track of this. Closing here.

janden commented on src/aspire/estimation/covar.py:
```diff
@@ -44,13 +44,14 @@ def compute_kernel(self):
         kernel = np.zeros((_2L, _2L, _2L, _2L, _2L, _2L), dtype=self.as_type)
-        filters_f = self.src.filters.evaluate_grid(L)
+        filters_f = self.src.eval_filter_grid(L)
         sq_filters_f = np.array(filters_f ** 2, dtype=self.as_type)
```
This ends up being quite expensive, since we're squaring each filter multiple times (due to the duplicates). The previous approach, while more complex, did avoid this issue (which could create memory and performance issues).

vineetbansal replied:
I've added an optional argument so we can specify the "power" of a filter's evaluation on a grid. This allows us to get away with squaring all the duplicate values.

Whether or not the method `eval_filter_grid` itself should exist is a separate issue #38. I suspect once we move to more general "transformers", we'll have a general approach to associate indices with transformers and group all operations that way.

janden commented on src/aspire/estimation/mean.py:
```diff
@@ -17,13 +17,14 @@ class MeanEstimator(Estimator):
         kernel = np.zeros((_2L, _2L, _2L), dtype=self.as_type)
-        filters_f = self.src.filters.evaluate_grid(self.L)
+        filters_f = self.src.eval_filter_grid(self.L)
         sq_filters_f = np.array(filters_f ** 2, dtype=self.as_type)
```
See comment in `covar.py`.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
-        ensure(data.ndim == 3, 'Initialize the image stack using a 3d ndarray.')
-        self._data = data
-        self.shape = self._data.shape
+        ensure(data.shape[0] == data.shape[1], 'Only square ndarrays are supported.')
```
I'm not quite sure how to address this, but we would really like the “image” axes (that is, x and y) to be the last two axes in the array, since NumPy is C-style (row-major) and we're likely going to be fed C-style arrays.

vineetbansal replied:
This is a good candidate for a single commit + PR. Opened issue #39 for this. Closing here.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
 
-    def __getitem__(self, item):
-        return self._data[item]
+        self.data = data.copy()
```
Why copy? Is it so bad if it shares the memory with some other object, as long as we're careful? These arrays could be rather large, so it would be good to reduce copying unless we have a good reason.

vineetbansal replied:
You're right .. no `copy` needed. Caller of constructor beware.
Fixed in 2b4369de05d23fc502c23aafb775862972bdb1c4

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+        return Image(self.data + other.data)
+
+    def __repr__(self):
+        return f'{self.n_images} images of size {self.res}x{self.res}'
```
If possible, replace `x` with `×` (U+00D7).

vineetbansal replied:
I'm always a bit leery of introducing non-ASCII characters unless absolutely necessary, especially for dumb Windows terminal programs that I mostly deal with. Let me close this here. We can revisit this at some future time.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+        return f'{self.n_images} images of size {self.res}x{self.res}'
+
+    def asnumpy(self):
+        warnings.warn('Access to underlying Image data is discouraged.'
```
Is it really necessary to warn here? There are perfectly legitimate reasons why one might want to do this.

vineetbansal replied:
Resolved in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+    def __repr__(self):
+        return f'{self.n_images} images of size {self.res}x{self.res}'
+
+    def asnumpy(self):
```
Is this standard naming for conversion to NumPy arrays? My only previous experience with this is in PyTorch and TensorFlow, which both use a `numpy()` method in their respective `Tensor` classes to return a NumPy array.

vineetbansal replied:
This was inspired from
https://docs-cupy.chainer.org/en/stable/reference/generated/cupy.asnumpy.html

I like the  `as` which makes it explicit, but I can be persuaded either way.

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+        if shifts.ndim == 1:
+            shifts = shifts[np.newaxis, :]
+
+        warnings.warn('Multiple implementations of im_translate. Running both for sanity check.'
```
Yeah we should really just pick one. It should be fairly straightforward to make sure they're doing the same thing. This isn't the place to run sanity checks.

vineetbansal replied:
Fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+        """
+        Downsample Image to a specific resolution. This method returns a new Image.
+        :param res: int - new resolution, should be <= the current resolution of this Image
+        :return: The down-sampled Image object.
```
“down-sampled” → “downsampled” (for consistency with “Downsample” and `downsample` above).

vineetbansal replied:
Fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/image/__init__.py:
```diff
@@ -42,92 +44,144 @@ def im_translate(im, shifts):
+
+        return Image(im_translated)
+
+    def downsample(self, res):
```
Maybe `res` → `ds_res` so as to not confuse it with `self.res`?

vineetbansal replied:
Fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/utils/filters.py:
```diff
@@ -136,6 +136,7 @@ def _evaluate(self, omega):
 
     def scale(self, c):
         self._scale *= c
+        return self
```
I don't understand this. Is `scale()` supposed to return a new (scaled) filter or modify itself? Currently, it seems to do both.

vineetbansal replied:
It's modifying itself, but also returning itself so that these calls can be chained:

```
f.scale(2).evaluate(omegas)
```

 `return self` is common to support such scenarios.

vineetbansal replied:
Not returning `self` to avoid this confusion.

janden commented on src/aspire/utils/filters.py:
```diff
@@ -211,6 +212,7 @@ def _evaluate(self, omega):
 
     def scale(self, c=1):
         self.pixel_size *= c
```
An `IdentityFilter`, being a `ScalarFilter`, remains invariant during scaling, so should just return `self` like its superclass (no need to define the method actually since it inherits that).

vineetbansal replied:
`IdentityFilter` is not overriding anything in it's superclass except the `value` (which is 1). The fragment above is from `CTFFilter`. Close if all good?

janden commented on tests/test_image.py:
```diff
@@ -0,0 +1,34 @@
+        im2 = _im_translate2(self.im_np, -shifts.reshape(2, 1))
+
+        self.assertTrue(np.allclose(im.asnumpy(), im1))
+        self.assertTrue(np.allclose(im1, im2))
```
Shouldn't we also test that this corresponds to an actual (manual) shift of the image?

vineetbansal replied:
Added in 9fdaaa75c42d64b3ce63915c7ef34ab3b9692035

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,110 +1,163 @@
-        self.filters = filters
-        self.indices = indices
+        :param L: resolution of (square) images (int)
+        :param n: The total no. of images available
```
Write out “number” (also below).

vineetbansal replied:
Fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,110 +1,163 @@
+
+    @filters.setter
+    def filters(self, values):
+        self.set_metadata('__filter', values)
```
This is a little confusing. What type of object do we assume `values` is, a list of filter objects? If it doesn't have the attributes listed below (for example, it's a `ScalarFilter` or something), it just puts `np.nan` everywhere?

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,110 +1,163 @@
+        :param values: A scalar or vector (of length |indices|) of replacement values.
+        :param indices: A list of 0-based indices indicating the indices for which to modify metadata.
+            If indices is None, then all indices in this Source object are modified. In this case,
+            values should either be a scalar or a vector of length equal to the total no. of images, |self.n|.
```
“no.” → “number”.

vineetbansal replied:
Fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,110 +1,163 @@
+        Get metadata field information of this ImageSource for selected indices
+        :param metadata_fields: A string, of list of strings, representing the metadata field(s) to be queried.
+        :param indices: A list of 0-based indices indicating the indices for which to get metadata.
+            If indices is None, then
```
Then what?

vineetbansal replied:
fixed.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -115,7 +168,31 @@ def _images(self, start=0, num=None):
 
         """
-        raise NotImplementedError('Subclasses should implement this and return an Image object')
+        raise NotImplementedError('Subclasses should implement this and return a 3d ndarray')
```
Why not an `Image` object? Also, “3d” → “3D”.

vineetbansal replied:
Changed in 7e3d482155a766fb98185c88f51d9e6b8313f786.
`ImageSource._images` method now expected to return an `Image` object.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -115,7 +168,31 @@ def _images(self, start=0, num=None):
-        raise NotImplementedError('Subclasses should implement this and return an Image object')
+        raise NotImplementedError('Subclasses should implement this and return a 3d ndarray')
+
+    def eval_filters(self, im_orig, start=0, num=None):
```
Wouldn't it be straightforward to have `num` default to `np.inf`?

I'm also not clear on what this function does. It takes a NumPy array, converts it to a set of `Image` objects, filters them, and converts back to NumPy. Why do we want to do this? It also doesn't fit with the name `eval_filters`, which suggests that we're calculating the filter values for some set of frequencies. As discussed below, it's also not clear that this should be a function of the `ImageSource` object itself, since it principally relates to the filters.

vineetbansal replied:
The whole point of `eval_filters` to me seems to be "ImageSource, apply your configured filters, if you will, not on your images, but on these images that I'm giving you (because they're part of an optimization loop and I'm trying to see how good my images are)".

As such, you're right that the Filter object should be responsible for doing this. However, we don't have a combined `Filter` object that takes the responsibility of multiple filters and their indices (now that `SourceFilter` has gone away). I'm opening a new issue #43  on this. My suspicion is that once we have a general-purpose `Transformer` class that maintains the association between transformation function <-> indices, that would start to clear up things.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -128,9 +205,9 @@ def images(self, start=0, num=None, apply_noise=False):
-            im = self._images(start, num)
+            im = Image(self._images(start, num))
 
         if apply_noise:
```
At no point should we be applying noise to images. If noise is added anywhere, it's in the `Simulation` class as it simulates noisy projection images.

vineetbansal replied:
Fixed in 7e3d482155a766fb98185c88f51d9e6b8313f786

janden commented on src/aspire/source/__init__.py:
```diff
@@ -152,19 +229,20 @@ def _noise_images(self, start=0, num=None, noise_seed=0, noise_filter=None):
-        return im
+        return Image(im)
 
     def set_max_resolution(self, max_L):
```
It's not really setting the maximum resolution as *downsampling* the images in the `ImageSource`. Below, the filter settings are changed, and so is `self.L`, but how do we guarantee the the subclasses respect this downsampling?

vineetbansal replied:
Agreed - this should be handled through Transformations. I've opened an issue #46 to keep track of this. Closing here.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -177,30 +255,22 @@ def whiten(self, whiten_filter=None):
         images = self.images()
         logger.debug("Applying whitening filter to all images and caching")
-        whitened_images = whiten_source_filter(images[:, :, :])
+        whitened_images = Image(images[:, :, :]).filter(whiten_filter)
```
This is very expensive. We're calculating *all* the images, filtering them, and storing them in the `ImageSource` object. If the entire dataset is a couple hundred gigabytes, we're in trouble.

vineetbansal replied:
Opened a new issue on this (#44). Closing here.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -236,9 +306,32 @@ def vol_forward(self, vol, start, num):
+    def to_starfile(self, mrcs_filename, block_name='', fields=None):
+        df = self._metadata.copy()
+        df['_image_name'] = pd.Series(['{0:05}@{1}'.format(i+1, mrcs_filename) for i in range(self.n)])
+        df = df.rename(self._metadata_aliases, axis=1)
```
Unless the aliases have been changed to correspond to the `_rln`, these STAR files won't be very useful. We're probably better off hardcoding those aliases in `to_starfile`.

vineetbansal replied:
This is a non-issue now (in 836810a2ef2d85aeb930aee40e25a2152e68a642) since we no longer have an intermediate `StarfileStack` class. Aliases are applied for `RelionStarfileStack`, so one will never end up with a STAR file with non-relion fields. In other words, we only support a Relion STAR file, nothing else.

janden commented on src/aspire/source/__init__.py:
```diff
@@ -1,110 +1,163 @@
+    # Optional renaming of metadata fields, used
+    # These are used ONLY during serialization/deserialization (to/from Starfiles, for example).
+    _metadata_aliases = {
+        '_image_name':  '_image_name',
```
Why all the leading underscores?

vineetbansal replied:
I think it's more straightforward to have leading underscore in metadata fields, since that's what all fields in the STAR file need to have. Not having underscores and introducing/eliminating them magically while serializing/deserializing would be somewhat arbitrary.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+            field_names = []  # current field names inside a loop
+            rows = []         # rows to add to current loop
+
+            pbar = tqdm()
```
Not sure that we want to be producing output (other than to a log) in these functions. It should be up to the end user what gets printed.

vineetbansal replied:
Removed `tqdm` while doing STAR file parsing.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+        # attributes here, we simply make them 'public' without a leading underscore.
+        self.loops = loops
+        self.name = name
+        self.metadata = metadata
```
Might be confusing to call this `metadata`, since the `metadata` attribute of an `ImageSource` refers to a loop. Might want to change one or the other.

vineetbansal replied:
Renamed to `properties`

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+class Starfile:
+
+    def __init__(self, starfile_path=None, blocks=None, loop=None, dataframe=None, block_name=''):
+
```
Remove space.

vineetbansal replied:
Fixed in d852ac9b2d05d48cf52f03d873c1abaae338fd9e

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+               all([l1 == l2 for l1, l2 in zip(self.loops, other.loops)])
+
+
+class StarfileLoop:
```
What does this class do other than to wrap a `DataFrame`?

vineetbansal replied:
That's actually pretty much it .. I do like the `StarfileLoop` class from the usability perspective. Should the underlying implementation change, the contract can be kept the same. Plus a better `repr` and a simplified constructor.

vineetbansal replied:
Removed `StarfileLoop`. Just using a `DataFrame` directly.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+            self.init_from_starfile(starfile_path)
+        elif blocks is not None:
+            self.init_from_blocks(blocks)
+        elif loop is not None:
```
Do we need these two additional ways of initializing the `Starfile`? We can already initialize a block from a loop and a loop from a dataframe.

vineetbansal replied:
Removed multiple ways of initializing a `Starfile` object. Kept `starfile_path` and `blocks` as per discussion.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+        return self.data.equals(other.data)
+
+
+class Starfile:
```
`Starfile` or `StarFile`? Second might be more standard.

vineetbansal replied:
Closing here but opening an issue on this.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+        logger.info(f'Parsing starfile at path {starfile_path}')
+
+        with open(starfile_path, 'r') as f:
+
```
Remove space.

vineetbansal replied:
Fixed in d852ac9b2d05d48cf52f03d873c1abaae338fd9e

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+                    assert len(tokens) == len(field_names), \
+                        f'Error at line {i}. Expected {len(field_names)} values, got {len(tokens)}.'
+
+                    # rows.append({k: v for k, v in zip(field_names, tokens)})
```
Remove.

vineetbansal replied:
gone

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+                # When in a 'loop', any blank line implies we break out of the loop
+                if not line:
+                    if in_loop:
+                        if rows:  # We have accumulated data for a loop
```
What happens if `in_loop` is true, but `rows` is empty? We just continue looking for more rows without exiting the loop?

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+                if not line:
+                    if in_loop:
+                        if rows:  # We have accumulated data for a loop
+                            loops.append(StarfileLoop(field_names, rows))
```
Per our discussion, I thought we decided that we can only have one loop per block?

vineetbansal replied:
STAR files support multiple loops per block, so leaving as-is, in case we need this in the future.

https://pubs.acs.org/doi/pdf/10.1021/ci00002a020?rand=5hf0myi3

For Relion data STAR files, a hardcoded value of `0` for both the block index and loop index is used, without exposing these parameters to the caller.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+                    # we're looking at a data row
+                    tokens = line.split()
+                    assert len(tokens) == len(field_names), \
+                        f'Error at line {i}. Expected {len(field_names)} values, got {len(tokens)}.'
```
May want to make this a warning (or allow a flag that prevents it from erroring). I've seen a few STAR files that have malformed lines. In the end, it doesn't always matter (for instance, if it's a data STAR file, having the wrong parameters for a single image makes little difference), so it can be convenient to ignore these errors sometimes.

vineetbansal replied:
Made STAR file parsing more tolerant to field cardinality. Extra tokens are ignored, missing tokens defaulted to blank values.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+
+    def __init__(self, starfile_path=None, blocks=None, loop=None, dataframe=None, block_name=''):
+
+        self.blocks = self.block_names = None
```
Why not an `OrderedDict`?

vineetbansal replied:
Why not indeed. Fixed.

janden commented on src/aspire/io/starfile.py:
```diff
@@ -0,0 +1,196 @@
+    def __repr__(self):
+        return f'Starfile with {len(self.blocks)} blocks'
+
+    def __getitem__(self, item):
```
Would it also make sense to have `__iter__` here? I guess we won't be traversing STAR files that way, though.

vineetbansal replied:
`__getitem__` (which is immensely useful here) and `__len__` (less so), by working together, automatically support iteration. I've added a unit test for this in d50b818eca647f3eb25c557b581e30d459c96b74

janden commented on src/aspire/source/relion.py:
```diff
@@ -119,26 +192,63 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
+        return f'Starfile ({self.n} images of size {self.L}x{self.L})'
+
+    def _images(self, start=0, num=None):
+
```
Remove space?

vineetbansal replied:
gone

janden commented on src/aspire/source/relion.py:
```diff
@@ -119,26 +192,63 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
+        groups = df.groupby('__mrc_filepath')
+        n_workers = min(n_workers, len(groups))
+
+        pbar = tqdm(total=self.n)
```
Again, should probably remove these.

vineetbansal replied:
gone

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+
+        df = df.rename(reverse_metadata_aliases, axis=1)
+
+        # TODO: Check behavior if this is a single mrc file (no '@')
```
Does that happen?

vineetbansal replied:
Apparently not, so comment now obliterated.

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
-    def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_files=False, max_rows=None):
+    @classmethod
+    def starfile2df(cls, filepath, block_index_or_name=0, loop_index=0, ignore_missing_files=False, max_rows=None):
+
```
Remove space.

vineetbansal replied:
Gone

janden commented on src/aspire/source/__init__.py:
```diff
@@ -236,9 +306,32 @@ def vol_forward(self, vol, start, num):
+        if fields is not None:
+            df = df.drop([str(col) for col in df.columns if col not in fields], axis=1)
+
+        return Starfile(dataframe=df, block_name=block_name)
```
How do we deal with all the radian → degree conversions that are performed in `starfile2df`? Where are they reversed?

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
 
-    def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_files=False, max_rows=None):
+    @classmethod
+    def starfile2df(cls, filepath, block_index_or_name=0, loop_index=0, ignore_missing_files=False, max_rows=None):
```
No need to specify `block_index_or_name` and `loop_index`. These are always zero.

vineetbansal replied:
Gone

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+
+        # Adding a full-filepath field to the Dataframe helps us save time later
+        # Note that os.path.join works as expected when the second argument is an absolute path itself
+        df['__mrc_filepath'] = df['__mrc_filename'].apply(lambda filename: os.path.join(dirpath, filename))
```
It won't always be the case that the STAR file is in the same path as the particle images. It might be good to specify it as an option (and have it default to `os.path.dirname(filepath)` if not provided).

vineetbansal replied:
Support added in c004dd29afa5d9ae4c552cd4855bb7fc447e8889

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+        df['__mrc_filepath'] = df['__mrc_filename'].apply(lambda filename: os.path.join(dirpath, filename))
+
+        # Helper function to check if a file exists
+        def file_exists(filepath):
```
Why is this necessary?

vineetbansal replied:
I've routinely used this in development so far, where I don't want to fetch all mrcs files, but just a small subset of them (and don't want to edit the STAR file either, since there is no inherent guarantee on the ordering of rows), but just direct things to do a best-effort approach.

The default value of `ignore_missing_files` is `False` of course.

vineetbansal replied:
`ignore_missing_files` gone. We've kept `max_rows` for testing purposes.

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+        def file_exists(filepath):
+            return os.path.exists(filepath)
+
+        if max_rows is not None:
```
Why is this necessary? Can't it just truncate to the first `max_rows` rows?

vineetbansal replied:
Simplified logic after removing `ignore_missing_files` parameter.

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+        n_missing = sum(missing)
+        if ignore_missing_files:
+            if n_missing > 0:
+                logger.info(f'Dropping {n_missing} rows with missing mrc files')
```
Not sure we want to do this. Sometimes, you may want to read in a STAR file just for the metadata, even if the images in the MRC files aren't there. In that case, you still want to keep the rows. Might be good to warn that there are missing files, though.

vineetbansal replied:
Ok. This is a good candidate for a new issue, since it might involve detailed discussions on what one might legitimately do with this information at the level of an `ImageSource` object.

vineetbansal replied:
Opened issue #41

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+logger = logging.getLogger(__name__)
+
+
+class RelionStarfileStack(ImageSource):
```
I think `RelionSource` would be a better name.

vineetbansal replied:
Changed in e5b22843f315d1690ba9bda2b7c2cf2af80bdcdb

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+                 ignore_missing_files=False, max_rows=None):
         """
-        Load a Relion starfile at given filepath
+        Load starfile at given filepath
```
“starfile” → “STAR file”

vineetbansal replied:
Fixed in e5b22843f315d1690ba9bda2b7c2cf2af80bdcdb

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
         """
-        Load a Relion starfile at given filepath
+        Load starfile at given filepath
         :param filepath: Absolute or relative path to .star file
```
“.star” → “STAR” (or change other occurrences of “STAR” to “.star”)

vineetbansal replied:
fixed in e5b22843f315d1690ba9bda2b7c2cf2af80bdcdb

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
         :param filepath: Absolute or relative path to .star file
         :param pixel_size: the pixel size of the images in angstroms (Default 1)
         :param B: the envelope decay of the CTF in inverse square angstrom (Default 0)
+        :param n_workers: No. of threads to spawn to read referenced .mrcs files (Default -1 to auto detect)
```
“No.” → “number”

vineetbansal replied:
fixed

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
         :param pixel_size: the pixel size of the images in angstroms (Default 1)
         :param B: the envelope decay of the CTF in inverse square angstrom (Default 0)
+        :param n_workers: No. of threads to spawn to read referenced .mrcs files (Default -1 to auto detect)
+        :param block_index_or_name: An integer specifying the block index (0-indexed), of a string specifying
```
Again, these parameters aren't needed.

vineetbansal replied:
gone

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+            the block name
+        :param loop_index: An integer specifying the loop index (0-indexed)
         :param ignore_missing_files: Whether to ignore missing MRC files or not (Default False)
         :param max_rows: Maximum no. of rows in .star file to read. If None (default), all rows are read.
```
“no.” → “number”

vineetbansal replied:
done

janden commented on src/aspire/source/relion.py:
```diff
@@ -1,84 +1,138 @@
+        :param loop_index: An integer specifying the loop index (0-indexed)
         :param ignore_missing_files: Whether to ignore missing MRC files or not (Default False)
         :param max_rows: Maximum no. of rows in .star file to read. If None (default), all rows are read.
             Note that this refers to the max no. of images to load, not the max. number of .mrcs files (which may be
```
“max” → “maximum”

vineetbansal replied:
done

janden commented on src/aspire/source/relion.py:
```diff
@@ -87,25 +141,44 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
-
-        StarfileStack.__init__(self, filepath, n_workers=n_workers, ignore_missing_files=ignore_missing_files,
-                               max_rows=max_rows)
+        logger.debug(f'Creating ImageSource from starfile at path {filepath}')
```
“starfile” → “STAR file”

vineetbansal replied:
done

janden commented on src/aspire/source/relion.py:
```diff
@@ -87,25 +141,44 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
+            raise RuntimeError('No mrcs files found for starfile!')
+
+        # Peek into the first image and populate some attributes
+        first_mrc_filepath = metadata.iloc[0]['__mrc_filepath']
```
Maybe the following lines (156–169) should be in a separate function?

vineetbansal replied:
Opened a new issue #47 on this. Closing here.

janden commented on src/aspire/source/relion.py:
```diff
@@ -119,26 +192,63 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
         )
+
+    def __str__(self):
+        return f'Starfile ({self.n} images of size {self.L}x{self.L})'
```
“Starfile” → “RelionStarfileStack” (or “RelionSource” if we want to do that)

vineetbansal replied:
Using `RelionSource` as the name of the class now.

janden commented on src/aspire/source/relion.py:
```diff
@@ -119,26 +192,63 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
+            arr = mrcfile.open(filepath).data
+            data = arr[df['__mrc_index'] - 1, :, :].T
+
+            if self.L < self._L:
```
This is confusing. What's `self.L` and what's `self._L`? Maybe need more descriptive names.

vineetbansal replied:
Yes. This is an issue that is best handled by the upcoming `Transformer` class. Opened an issue #46 to keep track of this. Closing here.

janden commented on src/aspire/source/relion.py:
```diff
@@ -119,26 +192,63 @@ def __init__(self, filepath, pixel_size=1, B=0, n_workers=-1, ignore_missing_fil
+        n_workers = min(n_workers, len(groups))
+
+        pbar = tqdm(total=self.n)
+        with futures.ThreadPoolExecutor(n_workers) as executor:
```
Is it necessary to multithread this? We typically won't be loading all that many images at a time (not more than a few thousand) to process them, so it's likely that we'd want to multithread that outer loop in many cases.

vineetbansal replied:
This multi-threading brought down the reading of a huge STAR file (and 330GB of associated mrcs files) from several hours down to 20 minutes. This might prove key once we start experimenting with real datasets. Better to leave it in there rather than having to add it later.

janden commented on src/aspire/source/simulation.py:
```diff
@@ -13,44 +13,43 @@
                 np.random.random(n) * 2 * np.pi
             ))
-        return angles
+        return angles * 180 / np.pi
```
Would be better to have them in radians.

vineetbansal replied:
All angles in radians now (except when dealing directly with metadata fields).

janden commented on src/aspire/source/simulation.py:
```diff
@@ -13,44 +13,43 @@
         self.C = C
         self.vols = self._gaussian_blob_vols(L=self.L, C=self.C, seed=0)
 
     def _uniform_random_angles(self, n, seed=None):
```
This is general enough to deserve a function outside of `Simulation`.

vineetbansal replied:
Fixed in an upcoming commit..

janden commented on src/aspire/source/simulation.py:
```diff
@@ -13,44 +13,43 @@
-            dtype=dtype
-        )
-
+        angles = angles or self._uniform_random_angles(n, seed=0)
```
We should allow the user to specify a seed for the simulation. Sometimes we'll want to run the same simulation multiple times, but with some randomness in each trial.

vineetbansal replied:
Fixed in cfc6e40895f49fe43a864c94ca75964e5d90a218

janden commented on src/aspire/source/simulation.py:
```diff
@@ -13,44 +13,43 @@
+        angles = angles or self._uniform_random_angles(n, seed=0)
+
+        self.states = states
+        if filters is not None:
```
How would a user specify the filters *and* the filter indices in advance? Currently, it would always resample them.

vineetbansal replied:
Right now there's no way to do this. I've opened up an issue #45 on this.

janden commented on src/aspire/source/simulation.py:
```diff
@@ -13,44 +13,43 @@
-        return angles
+        return angles * 180 / np.pi
 
     def _gaussian_blob_vols(self, L=8, C=2, K=16, alpha=1, seed=None):
```
Again, might deserve its own function outside of the class.

vineetbansal replied:
We should use the ufuncs `np.deg2rad` and `np.rad2deg` for these.

vineetbansal replied:
This function (and all others) only have to deal with radians. The conversion to/from degrees is only done when modifying/accessing the underlying metadata fields.

As long as all future code concerns itself with the `angles` and `rots` attributes of the `ImageSource` class and doesn't mess around with the property setter/getter, everything will work.

We're using numpy's conversion functions for this, which being ufuncs, should be the fastest implementation we can use.

janden commented on tests/test_simulation_metadata.py:
```diff
@@ -0,0 +1,72 @@
+    metadata_aliases = {'greeting': 'my_greeting'}
+
+
+class SimTestCase(TestCase):
```
Shouldn't we also test the constructor, image generation methods, etc.?

janden commented on tests/test_starfileio.py:
```diff
@@ -0,0 +1,62 @@
+class StarfileTestCase(TestCase):
+    def setUp(self):
+        with importlib_resources.path(aspire.data, 'sample.star') as path:
+            self.starfile= Starfile(path)
```
Space before `=`.

vineetbansal replied:
Fixed in c7487a965a7c89b178f68cc1f84210f8da74b964

janden reviewed (commented):
Most of this looks good, but there are a few things here and there that need to be addressed. Part of it is addressing older issues from the existing code, but we might as well look at it now.

vineetbansal committed:
Some low hanging fruits fixed from PR #35 review.

vineetbansal committed:
Support for evaluating radial filters on a 1d omega vector.

Merged changes from develop for easier merging later.

vineetbansal committed:
Merge remote-tracking branch 'origin/big_refactor' into big_refactor

vineetbansal committed:
Removed the filter_values parameter for the Image.filter method.

vineetbansal committed:
Not making a copy of underlying numpy data in Image class.

vineetbansal committed:
Ability of specify a seed for Simulation (which is then reused internally).

vineetbansal committed:
Removed leading blank line before any constructor.

vineetbansal committed:
Killed commented line, added a unit test on Starfile iteration.

vineetbansal committed:
Renamed RelionStarfileStack to RelionSource

vineetbansal committed:
Introduced an 'indices' argument to images() and _images() methods of an ImageSource.

For accessing images at non-contiguous indices, and explicit argument 'indices' (which, if not None, takes precedence over start and num) is introduced. This is typically left unused, but will be helpful in a forthcoming method 'groupby' for an ImageSource that will group Image objects by certain metadata fields.

vineetbansal committed:
Ability to get 'grouped' Image objects from an ImageSource, grouped by any metadata field.

vineetbansal committed:
Filter methods that modify the object no longer return self.

Chaining method calls usually leads to the expectation that the object remains unmodified, so disallowing chaining.

vineetbansal committed:
Grouping filters during calculation in ImageSource.eval_filter_grid

Also removed some reshapings in Fortran order (introduced during intial port from Matlab, and which cancel each other out before the function returns) for a couple of functions.

vineetbansal committed:
Eliminated StarfileLoop class

StarfileLoop was a thin wrapper around a DataFrame. Using DataFrame(s) directly inside StarfileBlock now.

vineetbansal committed:
Bug fix after removing StarfileLoop class.

vineetbansal committed:
Renamed 'metadata' attribute of a StarfileBlock to 'properties'.

This is to avoid confusion with the usage of 'metadata' as used for ImageSource class.

vineetbansal committed:
ignore_missing_files logic removed from RelionSource

ignore_missing_files not used anymore, simplifying code. Starfile parsing more resilient to missing/extra fields found for loop data. For RelionSource, block/loop index is always assumed 0. Other minor fixes.

vineetbansal committed:
Function to convert rotation matrices to angles. Used to set mutually-linked attributes 'angles' and 'rots' for an ImageSource object (backed by 3 metadata fields).

vineetbansal committed:
Fewer ways to initialize Starfile class.

Also moved the metadata_aliases one level up in the ImageSource hierarchy, so that saving Simulation objects as STAR files results in a valid Relion STAR file.

vineetbansal committed:
Renamed Starfile class to StarFile.

vineetbansal committed:
Using OrderedDict to keep track of StarFileBlock objects inside a StarFile class.

vineetbansal committed:
Storing rotation angles in degrees in metadata table

This is the only place where angles are stored in degrees. All external access through 'angles' and 'rots' attributes is in radians. No other function in entire ASPIRE codebase needs to support angles in degrees. Using np ufuncs for deg <-> rad conversion.

vineetbansal committed:
Ability to specify a 'power' for evaluation of a filter grid for a Source.

vineetbansal committed:
ImageSource '_images' method now expected to return an Image object.

The public-facing 'images' method of ImageSource passes on any additional args/kwargs to the subclass's _images method. Moved 'apply_noise' argument from ImageSource (which would never need to do that) to Simulation.

vineetbansal committed:
RelionSource now supports an optional argument 'data_folder' in constructor.

This is to override relative paths to .mrcs files referenced in the STAR file.

vineetbansal committed:
Updated comment.

vineetbansal committed:
Added a test to compare Image shift vs. a numpy roll.

vineetbansal committed:
Minor issues raised in PR #35 fixed.

Starfile object now saves given a file pointer, so that we can do incremental saves (in a pipelined scenario, for example).

vineetbansal committed:
Merge branch 'develop' into big_refactor
